hazelcast/src/main/java/com/hazelcast/multimap/impl/MultiMapContainerSupport.java

```diff
@@ -66,7 +72,146 @@ public MultiMapValue getMultiMapValueOrNull(Data dataKey) {
+            if (iterator.next().getRecordId() == recordId) {
+                iterator.remove();
+                decrSize(1);
+                if (coll.isEmpty()) {
+                    multiMapValues.remove(dataKey);
+                }
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public long removeValue(Data dataKey, MultiMapRecord record) {
+        MultiMapValue multiMapValue = getMultiMapValueOrNull(dataKey);
+        if (multiMapValue == null) {
+            return -1;
+        }
+        Collection<MultiMapRecord> coll = multiMapValue.getCollection(false);
+        Iterator<MultiMapRecord> iterator = coll.iterator();
+        while (iterator.hasNext()) {
+            MultiMapRecord multiMapRecord = iterator.next();
+            if (multiMapRecord.equals(record)) {
+                iterator.remove();
+                decrSize(1);
+                if (coll.isEmpty()) {
+                    multiMapValues.remove(dataKey);
+                }
+                return multiMapRecord.getRecordId();
+            }
+        }
+        return -1;
+    }
+
+    public List<MultiMapRecord> removeAllValues(Data dataKey, Collection<Long> recordIds) {
+        List<MultiMapRecord> removed = new LinkedList<MultiMapRecord>();
+        MultiMapValue multiMapValue = getOrCreateMultiMapValue(dataKey);
+        for (Long recordId : recordIds) {
+            if (!multiMapValue.containsRecordId(recordId)) {
+                return removed;
+            }
+        }
+
+        int numOfRecsRemoved = 0;
+        Collection<MultiMapRecord> coll = multiMapValue.getCollection(false);
+        for (Long recordId : recordIds) {
+            Iterator<MultiMapRecord> iterator = coll.iterator();
+            while (iterator.hasNext()) {
+                MultiMapRecord record = iterator.next();
+                if (record.getRecordId() == recordId) {
+                    iterator.remove();
+                    removed.add(record);
+                    ++numOfRecsRemoved;
+                    break;
+                }
+            }
+        }
+        decrSize(numOfRecsRemoved);
+        if (coll.isEmpty()) {
+            multiMapValues.remove(dataKey);
+        }
+        return removed;
+    }
+
+    public boolean delete(Data dataKey) {
+        MultiMapValue value = multiMapValues.remove(dataKey);
+        if (value != null) {
+            decrSize(value.size());
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    public Collection<MultiMapRecord> remove(Data dataKey, boolean copyOf) {
+        MultiMapValue multiMapValue = multiMapValues.remove(dataKey);
+        if (multiMapValue != null) {
+            decrSize(multiMapValue.size());
+            return multiMapValue.getCollection(copyOf);
+        } else {
+            return null;
+        }
+    }
+
     public MultiMapConfig getConfig() {
         return config;
     }
+
+    public void incSize(int by) {
```

Comment:
These are vary dangerous methods because they are public. This can lead to procedural programming where it is difficult to determine if the object is in a valid state. Can these be made private?